transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,200 @@
+            }
+            Future<Boolean> f = healthCheck.isHealthy(ch, key);
+            if (f.isDone()) {
+                notifyHealthCheck(f, ch, key, promise);
+            } else {
+                f.addListener(new GenericFutureListener<Future<? super Boolean>>() {
+                    @Override
+                    public void operationComplete(Future<? super Boolean> future) throws Exception {
+                        notifyHealthCheck(future, ch, key, promise);
+                    }
+                });
+            }
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void notifyHealthCheck(Future<? super Boolean> future, C ch, K key, Promise<C> promise) {
+        if (future.isSuccess()) {
+            handler.channelAcquired(ch, key);
+            promise.setSuccess(ch);
+        } else {
+            ch.close();
+            acquire(key, promise);
+        }
+    }
+
+    private void newChannel(
+            final K key, final Promise<C> promise) {
+        Bootstrap bs;
+        EventLoop loop = key.eventLoop();
+        if (loop != null) {
+            bs = bootstrap.clone(loop);
+        } else {
+            bs = bootstrap.clone();
+        }
+        ChannelFuture f = bs.attr(KEY, key).connect(key.remoteAddress());
+        if (f.isDone()) {
+            notifyConnect(f, key, promise);
+        } else {
+            f.addListener(new ChannelFutureListener() {
+                @SuppressWarnings("unchecked")
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    notifyConnect(future, key, promise);
+                }
+            });
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void notifyConnect(ChannelFuture future, ChannelPoolKey key, Promise<C> promise) {
+        if (future.isSuccess()) {
+            C ch = (C) future.channel();
+            ch.attr(KEY).set(key);
+            promise.setSuccess(ch);
+        } else {
+            promise.setFailure(future.cause());
+        }
+    }
+
+    @Override
+    public Future<Boolean> release(C channel) {
+        return release(channel, ImmediateEventExecutor.INSTANCE.<Boolean>newPromise());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Future<Boolean> release(C channel, Promise<Boolean> promise) {
+        checkNotNull(channel, "channel");
+        checkNotNull(promise, "promise");
+        try {
+            K key = (K) channel.attr(KEY).getAndSet(null);
+            if (key != null) {
+                Deque<C> channels = pool.get(key);
+                if (channels == null) {
+                    channels = newDeque();
+                    Deque<C> old = pool.putIfAbsent(key, channels);
+                    if (old != null) {
+                        channels = old;
+                    }
+                }
+
+                if (channels.add(channel)) {
```

Comment:
Yep... The idea of the check was to allow all kind of things which may even is something like send a "NOOP" message to the remote peer before reuse the channel. That's why I said it may be expensive. 

@Scottmitch thanks for bring this up :)